### PR TITLE
Refactor error usage from the store level up and add API helpers

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary
This PR updates the store layer to consistently return errors and empty results, moves errors to the `errors.go` file and adds API level helpers to be able to correctly interpret different errors and associate the correspondant API response.

### CODE REVIEW

File: server/api/api_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		t.Run(fmt.Sprintf("%s should be a %d code", tc.Name, tc.ResponseCode), func(t *testing.T) {
+			r := httptest.NewRequest(http.MethodGet, "/test", nil)
+			w := httptest.NewRecorder()
+
+			testApi.errorResponse(w, r, tc.Error)
+			res := w.Result()
+
+			require.Equal(t, tc.ResponseCode, res.StatusCode)
+			require.Equal(t, "application/json", res.Header.Get("Content-Type"))
+			b, rErr := ioutil.ReadAll(res.Body)
```
Comment (wiggin77) [inline server/api/api_test.go:64]: Deprecated :wink: :stuck_out_tongue:

File: server/api/api.go [hunk 2]
```diff
@@ -174,7 +166,41 @@ func (a *API) userIsGuest(userID string) (bool, error) {
 
 // Response helpers
 
-func (a *API) errorResponse(w http.ResponseWriter, api string, code int, message string, sourceError error) {
+func (a *API) errorResponse(w http.ResponseWriter, r *http.Request, err error) {
```
Comment (wiggin77) [inline server/api/api.go:169]: I like this simplification of errorResponse  :+1:

File: server/api/blocks.go [hunk 3]
```diff
@@ -463,31 +450,23 @@ func (a *API) handleUndeleteBlock(w http.ResponseWriter, r *http.Request) {
...
-	}
-	if block == nil {
-		a.errorResponse(w, r.URL.Path, http.StatusNotFound, "", nil)
+		a.errorResponse(w, r, err)
 		return
 	}
 
 	if board.ID != block.BoardID {
-		a.errorResponse(w, r.URL.Path, http.StatusNotFound, "", nil)
+		a.errorResponse(w, r, model.NewErrNotFound("block"))
```
Comment (wiggin77) [inline server/api/blocks.go:464]: Would prefer to have block id passed to ErrNotFound.

File: server/api/boards.go [hunk 4]
```diff
@@ -127,40 +125,40 @@ func (a *API) handleCreateBoard(w http.ResponseWriter, r *http.Request) {
...
 	if err != nil {
-		a.errorResponse(w, r.URL.Path, http.StatusInternalServerError, "", err)
+		a.errorResponse(w, r, err)
 		return
 	}
 
 	var newBoard *model.Board
 	if err = json.Unmarshal(requestBody, &newBoard); err != nil {
-		a.errorResponse(w, r.URL.Path, http.StatusBadRequest, "", err)
+		a.errorResponse(w, r, model.NewErrBadRequest(""))
```
Comment (wiggin77) [inline server/api/boards.go:134]: I think the Unmarshal error would be helpful for troubleshooting.  

```suggestion
		a.errorResponse(w, r, model.NewErrBadRequest(err.Error()))
```

File: server/api/boards.go [hunk 5]
```diff
@@ -653,27 +633,23 @@ func (a *API) handleGetBoardMetadata(w http.ResponseWriter, r *http.Request) {
...
-		return
-	}
 	if err != nil {
-		a.errorResponse(w, r.URL.Path, http.StatusInternalServerError, "", err)
+		a.errorResponse(w, r, err)
 		return
 	}
 	if board == nil || boardMetadata == nil {
-		a.errorResponse(w, r.URL.Path, http.StatusNotFound, "", nil)
+		a.errorResponse(w, r, model.NewErrNotFound(""))
```
Comment (wiggin77) [inline server/api/boards.go:641]: Include board ID?

File: server/api/boards_and_blocks.go [hunk 6]
```diff
@@ -229,59 +223,51 @@ func (a *API) handlePatchBoardsAndBlocks(w http.ResponseWriter, r *http.Request)
...
+			a.errorResponse(w, r, err2)
 			return
 		}
 
 		if teamID == "" {
 			teamID = board.TeamID
 		}
 		if teamID != board.TeamID {
-			a.errorResponse(w, r.URL.Path, http.StatusBadRequest, "", nil)
+			a.errorResponse(w, r, model.NewErrBadRequest(""))
```
Comment (wiggin77) [inline server/api/boards_and_blocks.go:252]:
```suggestion
			a.errorResponse(w, r, model.NewErrBadRequest("mismatched team ID"))
```

File: server/api/boards_and_blocks.go [hunk 7]
```diff
@@ -229,59 +223,51 @@ func (a *API) handlePatchBoardsAndBlocks(w http.ResponseWriter, r *http.Request)
...
-		}
-		if block == nil {
-			a.errorResponse(w, r.URL.Path, http.StatusBadRequest, "", nil)
+			a.errorResponse(w, r, err2)
 			return
 		}
 
 		if _, ok := boardIDMap[block.BoardID]; !ok {
-			a.errorResponse(w, r.URL.Path, http.StatusBadRequest, "", nil)
+			a.errorResponse(w, r, model.NewErrBadRequest(""))
```
Comment (wiggin77) [inline server/api/boards_and_blocks.go:265]:
```suggestion
			a.errorResponse(w, r, model.NewErrBadRequest("missing board ID " + block.BoardID))
```

File: server/api/members.go [hunk 8]
```diff
@@ -222,37 +216,33 @@ func (a *API) handleJoinBoard(w http.ResponseWriter, r *http.Request) {
 
 	userID := getUserID(r)
 	if userID == "" {
-		a.errorResponse(w, r.URL.Path, http.StatusBadRequest, "", nil)
+		a.errorResponse(w, r, model.NewErrBadRequest(""))
```
Comment (wiggin77) [inline server/api/members.go:219]:
```suggestion
		a.errorResponse(w, r, model.NewErrBadRequest("missing user ID"))
```

File: server/api/members.go [hunk 9]
```diff
@@ -323,24 +313,20 @@ func (a *API) handleLeaveBoard(w http.ResponseWriter, r *http.Request) {
...
```
Comment (wiggin77) [inline server/api/members.go:316]:
```suggestion
		a.errorResponse(w, r, model.NewErrBadRequest("missing user ID"))
```

File: server/api/subscriptions.go [hunk 10]
```diff
@@ -50,20 +51,19 @@ func (a *API) handleCreateSubscription(w http.ResponseWriter, r *http.Request) {
...
-	if err != nil {
-		a.errorResponse(w, r.URL.Path, http.StatusInternalServerError, "", err)
+	if err = json.Unmarshal(requestBody, &sub); err != nil {
+		a.errorResponse(w, r, err)
 		return
 	}
 
 	if err = sub.IsValid(); err != nil {
-		a.errorResponse(w, r.URL.Path, http.StatusBadRequest, "", err)
+		a.customErrorResponse(w, r.URL.Path, http.StatusBadRequest, "", err)
```
Comment (wiggin77) [inline server/api/subscriptions.go:66]:
```suggestion
		a.errorResponse(w, r, model.NewErrBadRequest(err.Error()))
```

File: server/api/users.go [hunk 11]
```diff
@@ -225,17 +226,17 @@ func (a *API) handleGetUser(w http.ResponseWriter, r *http.Request) {
 
 	canSeeUser, err := a.app.CanSeeUser(session.UserID, userID)
 	if err != nil {
-		a.errorResponse(w, r.URL.Path, http.StatusInternalServerError, "", err)
+		a.errorResponse(w, r, err)
 		return
 	}
 	if !canSeeUser {
-		a.errorResponse(w, r.URL.Path, http.StatusNotFound, "", nil)
+		a.errorResponse(w, r, model.NewErrNotFound(""))
```
Comment (wiggin77) [inline server/api/users.go:233]: Add user ID?

File: server/services/store/mattermostauthlayer/mattermostauthlayer.go [hunk 12]
```diff
@@ -492,7 +496,7 @@ func (s *MattermostAuthLayer) GetFileInfo(id string) (*mmModel.FileInfo, error)
 		var appErr *mmModel.AppError
 		if errors.As(err, &appErr) {
 			if appErr.StatusCode == http.StatusNotFound {
-				return nil, nil
+				return nil, model.NewErrNotFound("file info")
```
Comment (wiggin77) [inline server/services/store/mattermostauthlayer/mattermostauthlayer.go:499]: This should include the file ID?

File: server/services/store/sqlstore/board.go [hunk 13]
```diff
@@ -242,7 +234,7 @@ func (s *SQLStore) getBoardsFieldsByCondition(db sq.BaseRunner, fields []string,
 	}
 
 	if len(boards) == 0 {
-		return nil, sql.ErrNoRows
+		return nil, model.NewErrNotFound("board")
```
Comment (wiggin77) [inline server/services/store/sqlstore/board.go:237]: Include board ID?

File: server/services/store/sqlstore/board.go [hunk 14]
```diff
@@ -410,7 +411,7 @@ func (s *SQLStore) patchBoard(db sq.BaseRunner, boardID string, boardPatch *mode
 		return nil, err
 	}
 	if existingBoard == nil {
-		return nil, BoardNotFoundErr{boardID}
+		return nil, model.NewErrNotFound("board")
```
Comment (wiggin77) [inline server/services/store/sqlstore/board.go:414]: same

**Comment by wiggin77** [comment]
This is great @mgdelacroix .   Much cleaner, no more returning `nil,nil` and error handling much simplified.  I think ErrNotFound should include the resource ID for debugging - I've left a bunch of comments like that then stopped as I was just adding noise.

File: server/api/blocks.go [hunk 15]
```diff
@@ -133,27 +127,25 @@ func (a *API) handleGetBlocks(w http.ResponseWriter, r *http.Request) {
...
 		}
-		if block != nil {
-			if block.BoardID != boardID {
-				a.errorResponse(w, r.URL.Path, http.StatusNotFound, "", nil)
-				return
-			}
-
-			blocks = append(blocks, *block)
+		if block.BoardID != boardID {
+			a.errorResponse(w, r, model.NewErrNotFound("board"))
```
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply server/api/blocks.go]: Agree, fixed on [db85b38](https://github.com/mattermost/focalboard/pull/3792/commits/db85b38428363ada96685dcfa414306b2c43821d)

File: server/api/blocks.go (hunk 3 above)
Comment (mgdelacroix) [reply server/api/blocks.go]: Agree, not found errors have been addressed on https://github.com/mattermost/focalboard/pull/3792/commits/db85b38428363ada96685dcfa414306b2c43821d, and the other error types that were created without context on https://github.com/mattermost/focalboard/pull/3792/commits/8e27b1af33059b45fd9f6b295bcfee5347981b7c

Merged and adapted the new Card API endpoints on https://github.com/mattermost/focalboard/pull/3792/commits/8bb7d209e90d0bf32b81b57668d1535dc250ef86

Comment (wiggin77) [reply server/api/blocks.go]: Thanks for updating the card APIs!

**Review by wiggin77 — APPROVED** [review]
LGTM!  :+1:  :+1:

File: server/api/api.go [hunk 16]
```diff
@@ -133,7 +125,7 @@ func (a *API) panicHandler(next http.Handler) http.Handler {
 					mlog.String("stack", string(debug.Stack())),
 					mlog.String("uri", r.URL.Path),
 				)
-				a.errorResponse(w, r.URL.Path, http.StatusInternalServerError, "", nil)
+				a.customErrorResponse(w, r.URL.Path, http.StatusInternalServerError, "", nil)
```
Comment (jespino) [inline server/api/api.go:128]: makes sense to use a `errorResponse(w, r, errors.New('panic during request handling'))` or something like that?

File: server/api/auth.go [hunk 17]
```diff
@@ -185,30 +185,30 @@ func (a *API) handleRegister(w http.ResponseWriter, r *http.Request) {
 	//     schema:
 	//       "$ref": "#/definitions/ErrorResponse"
 	if a.MattermostAuth {
-		a.errorResponse(w, r.URL.Path, http.StatusNotImplemented, "not permitted in plugin mode", nil)
+		a.customErrorResponse(w, r.URL.Path, http.StatusNotImplemented, "not permitted in plugin mode", nil)
```
Comment (jespino) [inline server/api/auth.go:188]: Probably a new `NotImplemented` error would be good here.

File: server/api/auth.go [hunk 18]
```diff
@@ -57,30 +57,30 @@ func (a *API) handleLogin(w http.ResponseWriter, r *http.Request) {
...
 	if a.MattermostAuth {
-		a.errorResponse(w, r.URL.Path, http.StatusNotImplemented, "", nil)
+		a.customErrorResponse(w, r.URL.Path, http.StatusNotImplemented, "", nil)
 		return
 	}
 
 	if len(a.singleUserToken) > 0 {
 		// Not permitted in single-user mode
-		a.errorResponse(w, r.URL.Path, http.StatusUnauthorized, "not permitted in single-user mode", nil)
+		a.customErrorResponse(w, r.URL.Path, http.StatusUnauthorized, "not permitted in single-user mode", nil)
```
Comment (jespino) [inline server/api/auth.go:70]: Shouldn't it be an `Unauthorized` error?

File: server/api/auth.go (hunk 17 above)
Comment (mgdelacroix) [reply server/api/auth.go]: Addressed in [1753287](https://github.com/mattermost/focalboard/pull/3792/commits/1753287d742a6de4013d85a194b0fb9fdb2decc4), only one `customErrorResponse` remains for `request entity too large`

**Review by jespino — APPROVED** [review]
now LGTM. The linter is complaining, but other than that is ready for merge.

As an extra work you can remove the last usage of customErrorResponse and remove the whole function to avoid having the option in the code and enforcing properly creating specific errors. This can be done directly here or as a new PR later. Feel free to do it however you prefer.

**Pull request merged**